### PR TITLE
util/chunk: introduce Chunk to replace current datum slice Row.

### PR DESCRIPTION
**Pull request by coocood**
Chunk is an efficient data structure to store rows of data.
It will improve both performance and memory usage.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @hanfei1991 @tiancaiamao @XuHuaiyu @zz-jason PTAL

**Comment by shenli** [comment]
Any benchmark result？

**Comment by coocood** [comment]
@shenli 
The test contains benchmark code.

Write(Append) is about 7 ns/op, Read is about 2 ns/op.

**Comment by zz-jason** [comment]
@coocood how about using a column store inside a chunk ?

**Comment by coocood** [comment]
@zz-jason 
I've considered column store, it's best for write once, read multiple times scenario.

But the write performance is worse than row store.

In our execution flow, Chunk should be used in each executor, doing reading and writing.

The executor reads a chunk from children, process it, then write data to a new chunk, return to its parent.

So I think row store is a better choice.

**Comment by ngaut** [comment]
Well done. We can move step by step. For now, chunk is a better choice.

**Comment by iamxy** [comment]
/run-integration-compatibility-test tidb_test=connor/add-compatible-test

**Comment by coocood** [comment]
@zz-jason @ngaut PTAL

File: util/chunk/chunk.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+	c.nullCount = 0
+	c.nullBitmap = c.nullBitmap[:0]
+	c.offsets = c.offsets[:0]
+	c.data = c.data[:0]
+	c.ifaces = c.ifaces[:0]
+}
+
+func (c *column) isNull(rowIdx int) bool {
+	nullByte := c.nullBitmap[rowIdx/8]
+	return nullByte&(1<<(uint(rowIdx)&7)) == 0
```
Comment (zz-jason) [inline util/chunk/chunk.go]: `0` in `nullBitmap` represents null is a little confusing, how about make `1` represent null ?

**Comment by ngaut** [comment]
Cool, Are we using standard apache arrow format? Could you add a reference link to the memory row format of apache arrow?

File: util/chunk/chunk.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+	c.finishAppendFixed()
+}
+
+func (c *column) appendNameValue(name string, val uint64) {
+	var buf [8]byte
+	*(*uint64)(unsafe.Pointer(&buf[0])) = val
+	c.data = append(c.data, buf[:]...)
+	c.data = append(c.data, name...)
+	c.finishAppendVar()
+}
```
Comment (zz-jason) [inline util/chunk/chunk.go]: how about move `appendDuration` and `appendMyDecimal` together with the fixed length value append functions and move `appendNameValue` together with the variant length value append functions ?

File: util/chunk/chunk.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	c.length = 0
+	c.nullCount = 0
+	c.nullBitmap = c.nullBitmap[:0]
+	c.offsets = c.offsets[:0]
+	c.data = c.data[:0]
+	c.ifaces = c.ifaces[:0]
+}
+
+func (c *column) isNull(rowIdx int) bool {
+	nullByte := c.nullBitmap[rowIdx/8]
```
Comment (zz-jason) [inline util/chunk/chunk.go]: rowIdx >> 3

**Comment by coocood** [comment]
@ngaut 
For each component (bitmap, offsets, data) it's the same as Apache Arrow, but each component has its own buffer instead of sharing a single one so we can build chunk before we know how many rows going to be.

File: util/chunk/chunk.go (hunk 1 above)
Comment (coocood) [reply util/chunk/chunk.go]: I agree that use 1 to represent null is more efficient, but Apache Arrow uses 0 to represent null.

File: util/chunk/chunk.go (hunk 3 above)
Comment (coocood) [reply util/chunk/chunk.go]: For efficiency, the compiler does it automatically.
For readability, I think '/ 8' is more readable.

File: util/chunk/chunk.go [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+// Values are appended in compact format and can be directly accessed without decoding.
+// When the chunk is done processing, we can reuse the allocated memory by resetting it.
+type Chunk struct {
+	columns []*column
+}
+
+// AddFixedLenColumn adds a fixed length column with elemLen.
+func (c *Chunk) AddFixedLenColumn(elemLen int) {
+	c.columns = append(c.columns, &column{
+		elemBuf: make([]byte, elemLen),
```
Comment (zz-jason) [inline util/chunk/chunk.go]: how about allocate `data` slice with a capacity ?

**Comment by coocood** [comment]
@zz-jason PTAL

File: util/chunk/chunk_test.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	b.StartTimer()
+	var sum int64
+	for i := 0; i < b.N; i++ {
+		var v int64
+		for j := 0; j < 8192; j++ {
+			v, _ = rowChk.GetRow(j).GetInt64(0)
+		}
+		sum += v
+	}
+	fmt.Println(sum)
```
Comment (tiancaiamao) [inline util/chunk/chunk_test.go]: Debugging?

File: util/types/mydecimal.go [hunk 6]
```diff
@@ -159,6 +159,9 @@ func digitsToWords(digits int) int {
 	return (digits + digitsPerWord - 1) / digitsPerWord
 }
 
+// MyDecimalStructSize is the struct size of MyDecimal.
+const MyDecimalStructSize = 40
```
Comment (tiancaiamao) [inline util/types/mydecimal.go]: `unsafe.Sizeof(MyDecimal{})` ?

File: util/chunk/chunk_test.go (hunk 5 above)
Comment (coocood) [reply util/chunk/chunk_test.go]: No, this is used to prevent compiler optimization.

File: util/types/mydecimal.go (hunk 6 above)
Comment (coocood) [reply util/types/mydecimal.go]: It will be more clear how large `MyDecimal` is, and we should not change the struct of the `MyDecimal`.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ddorian** [comment]
I'm trying to understand a little the `chunk` from a high level overview. Looks like each row is stored as 1 key-value in tikv, but when a `rocksdb block` is cached in memory it's converted to small column store (the chunk/block only) right ?
So on disk each row is a tuple while on ram each block is a small column store ?
Or chunk used only for memstore ?

**Comment by tiancaiamao** [comment]
> So on disk each row is a tuple while on ram each block is a small column store ?
Exactly. @ddorian

**Pull request merged**